File: 2020/25/Modular.hs
module Modular where

import           ModularArithmetic

modulus :: Integer
modulus = 20201227

solve :: (Integer, Integer) -> Integer
solve (k1, k2) | r1 == r2  = r1
               | otherwise = error "Encryption keys do not match up"
 where
  l1 = discreteLog 7 k1 modulus
  l2 = discreteLog 7 k2 modulus
  r1 = modularExp k1 l2 modulus
  r2 = modularExp k2 l1 modulus

File: 2020/25/Rust/src/modular.rs
mod modular_arithmetic;

use crate::MODULUS;
use modular_arithmetic::{discrete_log, modular_exp};

pub fn solve(keys: &(u128, u128)) -> Option<u128> {
    let (key_1, key_2) = *keys;
    let loop_size_1 = discrete_log(&7, &key_1, &MODULUS).unwrap();
    let loop_size_2 = discrete_log(&7, &key_2, &MODULUS).unwrap();
    let result_1 = modular_exp(&key_1, &loop_size_2, &MODULUS);
    let result_2 = modular_exp(&key_2, &loop_size_1, &MODULUS);
    if result_1 == result_2 {
        Some(result_1)
    } else {
        None
    }
}
